src/_pytest/monkeypatch.py

```diff
@@ -246,9 +245,10 @@ def setattr(
         if raising and oldval is notset:
             raise AttributeError(f"{target!r} has no attribute {name!r}")
 
-        # avoid class descriptors like staticmethod/classmethod
-        if inspect.isclass(target):
-            oldval = target.__dict__.get(name, notset)
+        # Prevent `undo` from polluting `vars(target)` with an object that was not in it
+        # before monkeypatching, such as inherited attributes or the results of
+        # descriptor binding.
+        oldval = vars(target).get(name, notset)
```

Comment:
How does this interact with slots?

Reply:
Nm, i reread on a PC and it's clear now

Reply:
i did a little more research, using vars in general breaks preexisting support for slots

setattr/getattr for "classes was hacked up to work around the descriptor protocol

i strongly suspect we need to use something more tricky to handle all edge-cases

the evil trick here is that for a type object the old hack works, for a "data descriptor" like cached property something semi-smart can work as well

however for other more powerful descriptors, we may face a an endless dig hole of edge-cases


the roundabout fix is to use a combination of getattr_static and also getattr_static on the type

so that if the result of getattr_static is the same object as the one on the type, we ought to use notset 

also there ought to be a warning on unknown descriptors as potential followup

also we cannot use getattr befor anymore, as that triggers cached properties, which is not what one would want when doing setattr

this is quite a rabbit hole

Reply:
this one is a little "monster" wrt scope,
after all this is all about gnarly edgecases

im not sure how to split it up in a good way